Panel API and Supporting Features

ShrutiVellanki opened the pull request:
## Description

Closes #2697 

Adds a new API for the creation and management (styling, contents, positioning) of multiple Panels on a Map instance. 

***New Additions:*** 
- `\api\src\panel.ts`
- `src\content\samples\api\new-panel\`
- `\src\content\styles\api\panel.css`

***Changes to:*** 
- `\api\src\map.ts`
- `\src\app\api-loader.ts`

## Testing

Extensive testing instructions for Panel Features and potential error cases:  https://github.com/ShrutiVellanki/fgpv-vpgf/blob/PanelApi/src/content/samples/api/new-panel-tests/panel/test.md

Basic Panel to copy/paste into demo console (for best results when checking style, test in index-mobile):
```js
        //creating Panel + opening
        let panel1 = RZ.mapInstances[0].createPanel('panel1');
        panel1.id;
        panel1.setPosition([10, 1], [13, 10]);
        panel1.open();

        //creating PanelElems
        let panelElem1 = panel1.createPanelElem("Layers");
        let panelElem2 = panel1.createPanelElem($("<div style='color: lightslategray'>Contents:</div>"));
        let panelElem3 = panel1.createPanelElem($.parseHTML('<input type="text" value="Search..." id="coolInput"></input>'));
        let panelElem4 = panel1.createPanelElem($('<p style="color: lightslategray">Controls:</p>'))
        let imgElem = panel1.createPanelElem($("<img id='coolImg' src='http://www.geographicguide.com/planet/images/mercator.jpg'></img>"));

        let btn = panel1.createBtn();
        btn.text = "Btn.";
        $(btn.element).click(function () {
            alert('Btn element clicked!')
        });

        //setting up the SVG icon

        let svg = $.parseHTML('<svg style="width:24px;height:24px" viewBox="0 0 24 24"><path fill="#ffffff" d="M17.9,17.39C17.64,16.59 16.89,16 16,16H15V13A1,1 0 0,0 14,12H8V10H10A1,1 0 0,0 11,9V7H13A2,2 0 0,0 15,5V4.59C17.93,5.77 20,8.64 20,12C20,14.08 19.2,15.97 17.9,17.39M11,19.93C7.05,19.44 4,16.08 4,12C4,11.38 4.08,10.78 4.21,10.21L9,15V16A2,2 0 0,0 11,18M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2Z"/></svg>');

        //assigning it to the btn2

        let btn2 = panel1.createBtn();
        btn2.icon = svg[0];

        //content setup
        let closeBtn = panel1.createPanelElem('x');
        $(panelElem2.element).append($("<br>"));
        $(panelElem2.element).append($("<br>"));
        $(panelElem2.element).append(btn.element);
        $(panelElem2.element).append(btn2.element);
        $(panelElem2.element).append($("<br>"));
        $(panelElem2.element).append($("<br>"));
        $(panelElem2.element).append(imgElem.element);

        //assigning contents
        panel1.content = panelElem2;
        panel1.content;

        //assinging controls
        panel1.controls = [closeBtn, panel1.createPanelElem('|'), panel1.createPanelElem('T'), panelElem1, 
        panel1.createPanelElem($('<br>')), panelElem4, panelElem3];
        panel1.controls;
```

## Documentation
- API docs: https://shrutivellanki.github.io/  (navigate to `api/src/panel` in sidebar)
- Inline comments

## Checklist
<!-- Quick checklist for items that are easy to miss -->

- [x] Commit messages follow [the guidelines](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/master/CONTRIBUTING.md#-git-commit-guidelines)
- [x] Release notes have been updated
- [x] PR targets the correct release version
- [x] Help files and documentation have been updated

Remember, it is a *muffin offence* to open a PR with any of the above checklist items incomplete.

Please keep the original issue up to date with the final solution, expected behaviour, and any additional notes for testers

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804)
<!-- Reviewable:end -->

milesap commented:
Please remove js/ts auto-generated docs

---

Review status: 0 of 317 files reviewed, all discussions resolved (waiting on @milesap, @barryytm, @james-rae, @dane-thomas, and @JahidAhmed)

---



*Comments from [Reviewable](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-:-LFbQcIm6TD6US-Bb850:bxf7khw)*
<!-- Sent from Reviewable.io -->

dane-thomas commented:
Review status: 0 of 10 files reviewed, 3 unresolved discussions (waiting on @ShrutiVellanki, @milesap, @barryytm, @james-rae, @dane-thomas, and @JahidAhmed)

---

*[api/src/map.ts, line 49 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFbj5uy7VAUkn1zj1gm:-LFbj5uy7VAUkn1zj1gn:bn86s7i) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/7c03f6d61ea52cd1bfd6015fc2b8296ef6d6bab1/api/src/map.ts#L49)):*
> ```XML
>     private _id: string;
>     //private _width: number;
>     //private _height: number;
> ```

Are these needed?

---

*[api/src/map.ts, line 61 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFbis4GEzKPUR-CRywd:-LFbis4GEzKPUR-CRywe:b-v5yx0n) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/7c03f6d61ea52cd1bfd6015fc2b8296ef6d6bab1/api/src/map.ts#L61)):*
> ```XML
>     private _panel_registry: Panel[];
> 
>    
> ```

Spaces; Also was this comment meant to be deleted?

---

*[api/src/map.ts, line 62 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFbiyBUB0BbUrBbj98O:-LFbiyBUB0BbUrBbj98P:b-f6q3mm) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/8c0ef7fcdd3197222fb74f1596d932f67f33c17d/api/src/map.ts#L62)):*
> ```XML
>         this._layers = new LayerGroup(this);
> 
>         // config set implies viewer loading via API
> ```

This one too?

---


*Comments from [Reviewable](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804)*
<!-- Sent from Reviewable.io -->

ShrutiVellanki commented:
Review status: 0 of 10 files reviewed, 3 unresolved discussions (waiting on @ShrutiVellanki, @milesap, @barryytm, @james-rae, @dane-thomas, and @JahidAhmed)

---

*[api/src/map.ts, line 49 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFbj5uy7VAUkn1zj1gm:-LFbqvcGAla00Ir3vNvG:bvcm4ta) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/7c03f6d61ea52cd1bfd6015fc2b8296ef6d6bab1/api/src/map.ts#L49)):*
<details><summary><i>Previously, dane-thomas (Dane Thomas) wrote…</i></summary><blockquote>

Are these needed?
</blockquote></details>

comments removed; not pushing until PRs get pulled 

---

*[api/src/map.ts, line 61 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFbis4GEzKPUR-CRywd:-LFbrSCd7XDaIBlszkPv:bym4wae) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/7c03f6d61ea52cd1bfd6015fc2b8296ef6d6bab1/api/src/map.ts#L61)):*
<details><summary><i>Previously, dane-thomas (Dane Thomas) wrote…</i></summary><blockquote>

Spaces; Also was this comment meant to be deleted?
</blockquote></details>

comment added back; not pushing until PRs get pulled

---

*[api/src/map.ts, line 62 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFbiyBUB0BbUrBbj98O:-LFbrRCb9OlA4r5RW0WV:bym4wae) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/8c0ef7fcdd3197222fb74f1596d932f67f33c17d/api/src/map.ts#L62)):*
<details><summary><i>Previously, dane-thomas (Dane Thomas) wrote…</i></summary><blockquote>

This one too?
</blockquote></details>

comment added back; not pushing until PRs get pulled

---


*Comments from [Reviewable](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804)*
<!-- Sent from Reviewable.io -->

milesap commented:
I think it would help if you also accept `(x, y)` tuples for `setPostion` and `setMinPosition` so that `setPosition((5,5), (10,15))` is also valid. The current way is fine but requires additional thinking and math :snail: 

I tested with two panels

#### Panel 1
`setPosition(0,210);` topLeft as 0,0 and bottomRight as 10,10
`setMinPosition(0,105);` topLeft as 0,0 and bottomRight as 5,5

#### Panel 2
`setPosition(6,110);` topLeft as 6,0 and bottomRight as 10,10

When I try and open panel 2 the `Uncaught Exception: conflicting panels, this panel cannot shrink any further to accomodate, panel closing.` is thrown. I was expecting Panel 1 to shrink horizontally from 0-10 to 0-6 to make room for the 6 to 10 wide panel 2. Let me know if I'm misunderstanding something

---

Reviewed 7 of 10 files at r1.
Review status: 7 of 10 files reviewed, 10 unresolved discussions (waiting on @dane-thomas, @ShrutiVellanki, @milesap, @barryytm, @james-rae, and @JahidAhmed)

---

*[api/src/map.ts, line 49 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFbipUB8whGuaB4y912:-LFbipUB8whGuaB4y913:byp18fd) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/7c03f6d61ea52cd1bfd6015fc2b8296ef6d6bab1/api/src/map.ts#L49)):*
> ```XML
>     private _id: string;
>     //private _width: number;
>     //private _height: number;
> ```

Remove

---

*[api/src/map.ts, line 57 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFbkp0rDPQiklBidVz1:-LFbkp0rDPQiklBidVz2:bbt3lnv) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/7c03f6d61ea52cd1bfd6015fc2b8296ef6d6bab1/api/src/map.ts#L57)):*
> ```XML
>     private _simpleLayer: SimpleLayer;
>     private _legendStructure: LegendStructure;
>     private _map_grid: number[][];
> ```

A suggestion: move these properties to the map prototype and avoid using starting `_`'s :

```js
export interface Map {
    mapGrid: number[][];
    innerShell: HTMLElement;
    panelRegistry: Panel[];
}
```

Then you can remove these. This will keep the object more organised in dev consoles and will improve the readability of the map class.

---

*[api/src/map.ts, line 67 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFbo6cADr4t3E4Tr6v6:-LFbo6cADr4t3E4Tr6v7:blypksi) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/7c03f6d61ea52cd1bfd6015fc2b8296ef6d6bab1/api/src/map.ts#L67)):*
> ```XML
>         this._ui = new UI(this);
>         this._layers = new LayerGroup(this);
>         this._inner_shell = <HTMLElement>document.getElementsByClassName('rv-inner-shell')[0];
> ```

What happens if there is more than one map on the page? Looks like it always uses the same element (the first map in DOM order).

---

*[api/src/panel.ts, line 2 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFbrh2gBSUE6mzOqvlM:-LFbrh2gBSUE6mzOqvlN:b-ugfcol) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/7c03f6d61ea52cd1bfd6015fc2b8296ef6d6bab1/api/src/panel.ts#L2)):*
> ```XML
> import { Observable } from 'rxjs';
> import { Subject } from 'rxjs';
> ```

Merge into `import { Observable, Subject } from 'rxjs';`

---

*[api/src/panel.ts, line 63 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFbrsSc28jUlkVymVAe:-LFbrsSc28jUlkVymVAf:bnfadn1) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/7c03f6d61ea52cd1bfd6015fc2b8296ef6d6bab1/api/src/panel.ts#L63)):*
> ```XML
>     private _heightChanged: Subject<any> = new Subject();
> 
>     private _panel_positions: PanelPositions;
> ```

Move as many of these into the objects prototype (and remove starting `_`)

---

*[api/src/panel.ts, line 116 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFbsIQU8phfsrc4Usyr:-LFbsIQU8phfsrc4Usys:ba3dfna) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/7c03f6d61ea52cd1bfd6015fc2b8296ef6d6bab1/api/src/panel.ts#L116)):*
> ```XML
>     * To be removed when API is finished. 
>     */
>     /*observableSubscribe(): void {
> ```

Remove?

---

*[src/app/api-loader.ts, line 44 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFbom4o0hBGmi-O84yj:-LFbom4o0hBGmi-O84yk:b5sw343) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/7c03f6d61ea52cd1bfd6015fc2b8296ef6d6bab1/src/app/api-loader.ts#L44)):*
> ```XML
>     get Panel(): typeof Panel { return Panel; }
>     get PanelElem(): typeof PanelElem { return PanelElem; }
>     get Btn(): typeof Btn { return Btn; }
> ```

Since panels and their elements are closely tied to map instances, these classes should be accessible via a map instance (i.e. `RZ.mapInstances[0].Panel` instead of the global `RZ.Panel`. Then `PanelElem` and `Btn` should be accessible via a panel instance since they are tied to a specific panel. These could be statically accessed or only through an instance, whichever you think is best.

---


*Comments from [Reviewable](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-:-LFc9MSgC0riLNWjq0Bo:be2gche)*
<!-- Sent from Reviewable.io -->

milesap commented:
Panel 2 should be `setPosition(6,210);` with same error

---

Review status: 7 of 10 files reviewed, 10 unresolved discussions (waiting on @dane-thomas, @ShrutiVellanki, @milesap, @barryytm, @james-rae, and @JahidAhmed)

---



*Comments from [Reviewable](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-:-LFcDsXM3tw4ApZQYSPC:b2wbzx5)*
<!-- Sent from Reviewable.io -->

barryytm commented:
Review status: 7 of 10 files reviewed, 22 unresolved discussions (waiting on @dane-thomas, @ShrutiVellanki, @milesap, @barryytm, @james-rae, and @JahidAhmed)

---

*[api/src/map.ts, line 49 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFbj5uy7VAUkn1zj1gm:-LFcBNaA697CI9jPimuY:b-4yqer3) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/7c03f6d61ea52cd1bfd6015fc2b8296ef6d6bab1/api/src/map.ts#L49)):*
<details><summary><i>Previously, ShrutiVellanki (Shruti Vellanki) wrote…</i></summary><blockquote>

comments removed; not pushing until PRs get pulled 
</blockquote></details>

You can still push what you have right for us to review.  Rebase will be needed regardless.

---

*[api/src/map.ts, line 71 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFcB8x883yJf1ef-Ifc:-LFcB8x883yJf1ef-Ifd:b-80x5a7) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/7c03f6d61ea52cd1bfd6015fc2b8296ef6d6bab1/api/src/map.ts#L71)):*
> ```XML
> 
>         if (config) {
> ```

Space

---

*[api/src/panel.ts, line 136 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFcETi74vUCzIVjr7mC:-LFcETi74vUCzIVjr7mD:bm9ssle) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/7c03f6d61ea52cd1bfd6015fc2b8296ef6d6bab1/api/src/panel.ts#L136)):*
> ```XML
>     * @private
>     */
>     private createPanelComponents(): void {
> ```

Perhaps create some sub helper functions for easier read?

---

*[api/src/panel.ts, line 310 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFcFkqh69jVcpVdTF0O:-LFcFRzi8KLIvs5TCLGA:b-kwa2zu) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/7c03f6d61ea52cd1bfd6015fc2b8296ef6d6bab1/api/src/panel.ts#L310)):*
> ```XML
> 
>                     //initialize attributes
>                     let minusSVG = $.parseHTML('<svg style="width:24px;height:24px" viewBox="0 0 24 24"><path fill="#ffffff" d="M19,13H5V11H19V13Z"/></svg>')[0];
> ```

Put the svg code into a file or a constant?

---

*[api/src/panel.ts, line 611 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFcFCxj3X7wpvnPQKrG:-LFcFCxj3X7wpvnPQKrH:b-lpn7qq) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/7c03f6d61ea52cd1bfd6015fc2b8296ef6d6bab1/api/src/panel.ts#L611)):*
> ```XML
>  * Shortcuts can be used to create special `PanelElem`s inteded for use on panel controls. There are four shortcuts:
>  * ```js
>  * let panelElem1 = new RZ.PanelElem('Layers');   //Title shortcut
> ```

Please remove if not needed or put a comment explaining why they were commented out.

---

*[api/src/panel.ts, line 642 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFcFbBN4cLZV9ulFNZq:-LFcFbBN4cLZV9ulFNZr:brdy8ye) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/7c03f6d61ea52cd1bfd6015fc2b8296ef6d6bab1/api/src/panel.ts#L642)):*
> ```XML
>     setElement(element?: string | HTMLElement | JQuery<HTMLElement>): void {
> 
>         let minusSVG = $.parseHTML('<svg style="width:24px;height:24px" viewBox="0 0 24 24"><path fill="#ffffff" d="M19,13H5V11H19V13Z"/></svg>')[0];
> ```

Same as before.

---

*[api/src/panel.ts, line 729 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFcFuyGBQ7M-2BeHvmN:-LFcFuyGBQ7M-2BeHvmO:b-30m3oj) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/7c03f6d61ea52cd1bfd6015fc2b8296ef6d6bab1/api/src/panel.ts#L729)):*
> ```XML
>  * 
>  * //an SVG Btn
>  * let svg = $.parseHTML('<svg style="width:24px;height:24px" viewBox="0 0 24 24"><path fill="#ffffff" d="M17.9,17.39C17.64,16.59 16.89,16 16,16H15V13A1,1 0 0,0 14,12H8V10H10A1,1 0 0,0 11,9V7H13A2,2 0 0,0 15,5V4.59C17.93,5.77 20,8.64 20,12C20,14.08 19.2,15.97 17.9,17.39M11,19.93C7.05,19.44 4,16.08 4,12C4,11.38 4.08,10.78 4.21,10.21L9,15V16A2,2 0 0,0 11,18M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2Z"/></svg>');
> ```

remove or explain.

---

*[src/app/app-loader.js, line 40 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFcBxKVDUbTHBKej3O1:-LFcBxKVDUbTHBKej3O2:bxewgfu) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/7c03f6d61ea52cd1bfd6015fc2b8296ef6d6bab1/src/app/app-loader.js#L40)):*
> ```JavaScript
> import '../content/styles/main.scss';
> import './ui/fancyHover';
> import '../content/styles/api/panel.css'
> ```

Please insert a newline at the end of the file.

---

*[src/content/samples/api/new-panel-tests/panel/panel.css, line 97 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFcCjIK0VuYwtUgXj0i:-LFcC3UaEHw8ydt2VjJf:bj0fxw1) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/7c03f6d61ea52cd1bfd6015fc2b8296ef6d6bab1/src/content/samples/api/new-panel-tests/panel/panel.css#L97)):*
> ```CSS
>     display: inline-block;
>     padding: 5px;
> }
> ```

New line.Newline, please.

---

*[src/content/samples/api/new-panel-tests/panel/panel-test.html, line 16 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFcCGeB73dDLwLd7ExT:-LFcCGeCE08AaZcjwXpO:bc4dfvl) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/7c03f6d61ea52cd1bfd6015fc2b8296ef6d6bab1/src/content/samples/api/new-panel-tests/panel/panel-test.html#L16)):*
> ```HTML
>     <script src="panel-test.js"></script>
> </body>
> </html>
> ```

Newline, please.

---

*[src/content/samples/api/new-panel-tests/panel/panel-test.js, line 63 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFcCm1MCtY3z42n2Vqh:-LFcCm1MCtY3z42n2Vqi:b-65tnpx) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/7c03f6d61ea52cd1bfd6015fc2b8296ef6d6bab1/src/content/samples/api/new-panel-tests/panel/panel-test.js#L63)):*
> ```JavaScript
>         RZ.mapInstances[0].panelRegistry[0].controls;
> 
>         //RZ.mapInstances[0].panelRegistry[0].width = 300;
> ```

Remove if not needed or put a comment to specify why you have commented them out.

---

*[src/content/samples/api/new-panel-tests/panel/panel-test.js, line 66 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFcCL2K4T0zadV7c_Ni:-LFcCL2K4T0zadV7c_Nj:bc4dfvl) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/7c03f6d61ea52cd1bfd6015fc2b8296ef6d6bab1/src/content/samples/api/new-panel-tests/panel/panel-test.js#L66)):*
> ```JavaScript
>         //RZ.mapInstances[0].panelRegistry[0].height = 400;*/
>     });
> });
> ```

Newline, please.

---

*[src/content/styles/api/panel.css, line 97 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFcChxg1VTz7ekcCMu7:-LFcChxg1VTz7ekcCMu8:bc4dfvl) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/7c03f6d61ea52cd1bfd6015fc2b8296ef6d6bab1/src/content/styles/api/panel.css#L97)):*
> ```CSS
>     display: inline-block;
>     padding: 5px;
> }
> ```

Newline, please.

---


*Comments from [Reviewable](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804)*
<!-- Sent from Reviewable.io -->

ShrutiVellanki commented:
Review status: 1 of 8 files reviewed, 22 unresolved discussions (waiting on @milesap, @dane-thomas, @ShrutiVellanki, @barryytm, @james-rae, and @JahidAhmed)

---

*[api/src/map.ts, line 49 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFbipUB8whGuaB4y912:-LFcYq4S1L4YJwfEnzh-:b-896fix) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/7c03f6d61ea52cd1bfd6015fc2b8296ef6d6bab1/api/src/map.ts#L49)):*
<details><summary><i>Previously, milesap (Miles Petrov) wrote…</i></summary><blockquote>

Remove
</blockquote></details>

Done.

---

*[api/src/map.ts, line 49 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFbj5uy7VAUkn1zj1gm:-LFcfpWa1uRW9uJD721E:b-896fix) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/7c03f6d61ea52cd1bfd6015fc2b8296ef6d6bab1/api/src/map.ts#L49)):*
<details><summary><i>Previously, barryytm (Barry Yung) wrote…</i></summary><blockquote>

You can still push what you have right for us to review.  Rebase will be needed regardless.
</blockquote></details>

Done.

---

*[api/src/map.ts, line 71 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFcB8x883yJf1ef-Ifc:-LFcYGLJDuGgHtTqM3dP:b-896fix) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/7c03f6d61ea52cd1bfd6015fc2b8296ef6d6bab1/api/src/map.ts#L71)):*
<details><summary><i>Previously, barryytm (Barry Yung) wrote…</i></summary><blockquote>

Space
</blockquote></details>

Done.

---

*[api/src/panel.ts, line 2 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFbrh2gBSUE6mzOqvlM:-LFcXybX-m67fl8yJ0Ro:b-896fix) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/7c03f6d61ea52cd1bfd6015fc2b8296ef6d6bab1/api/src/panel.ts#L2)):*
<details><summary><i>Previously, milesap (Miles Petrov) wrote…</i></summary><blockquote>

Merge into `import { Observable, Subject } from 'rxjs';`
</blockquote></details>

Done.

---

*[api/src/panel.ts, line 116 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFbsIQU8phfsrc4Usyr:-LFcXqugCJPpS7flLC-D:b-896fix) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/7c03f6d61ea52cd1bfd6015fc2b8296ef6d6bab1/api/src/panel.ts#L116)):*
<details><summary><i>Previously, milesap (Miles Petrov) wrote…</i></summary><blockquote>

Remove?
</blockquote></details>

Done.

---

*[api/src/panel.ts, line 136 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFcETi74vUCzIVjr7mC:-LFcXB-68X4kGuvfd6OT:be306bi) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/7c03f6d61ea52cd1bfd6015fc2b8296ef6d6bab1/api/src/panel.ts#L136)):*
<details><summary><i>Previously, barryytm (Barry Yung) wrote…</i></summary><blockquote>

Perhaps create some sub helper functions for easier read?
</blockquote></details>

Done. (created `windowResize` helper)

---

*[api/src/panel.ts, line 611 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFcFCxj3X7wpvnPQKrG:-LFcVG7k2bM7L_CN5WjK:bnfsjkz) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/7c03f6d61ea52cd1bfd6015fc2b8296ef6d6bab1/api/src/panel.ts#L611)):*
<details><summary><i>Previously, barryytm (Barry Yung) wrote…</i></summary><blockquote>

Please remove if not needed or put a comment explaining why they were commented out.
</blockquote></details>

Part of the documentation for the `PanelElem` class. (see doc pages)

---

*[api/src/panel.ts, line 729 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFcFuyGBQ7M-2BeHvmN:-LFcU4QS9eWNTZ2gaveq:b9e3c7c) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/7c03f6d61ea52cd1bfd6015fc2b8296ef6d6bab1/api/src/panel.ts#L729)):*
<details><summary><i>Previously, barryytm (Barry Yung) wrote…</i></summary><blockquote>

remove or explain.
</blockquote></details>

Part of the documentation for the `Btn` class.  Explained the SVG setup with a clearer comment.

---

*[src/content/samples/api/new-panel-tests/panel/panel-test.html, line 16 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFcCGeB73dDLwLd7ExT:-LFcRADHBVacsuzpYTmu:b-896fix) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/7c03f6d61ea52cd1bfd6015fc2b8296ef6d6bab1/src/content/samples/api/new-panel-tests/panel/panel-test.html#L16)):*
<details><summary><i>Previously, barryytm (Barry Yung) wrote…</i></summary><blockquote>

Newline, please.
</blockquote></details>

Done.

---

*[src/content/samples/api/new-panel-tests/panel/panel-test.js, line 63 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFcCm1MCtY3z42n2Vqh:-LFcQ1ocAKSZT85wa8pV:b-896fix) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/7c03f6d61ea52cd1bfd6015fc2b8296ef6d6bab1/src/content/samples/api/new-panel-tests/panel/panel-test.js#L63)):*
<details><summary><i>Previously, barryytm (Barry Yung) wrote…</i></summary><blockquote>

Remove if not needed or put a comment to specify why you have commented them out.
</blockquote></details>

Done.

---

*[src/content/samples/api/new-panel-tests/panel/panel-test.js, line 66 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFcCL2K4T0zadV7c_Ni:-LFcPv40B5nV-wrtKeRN:b-896fix) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/7c03f6d61ea52cd1bfd6015fc2b8296ef6d6bab1/src/content/samples/api/new-panel-tests/panel/panel-test.js#L66)):*
<details><summary><i>Previously, barryytm (Barry Yung) wrote…</i></summary><blockquote>

Newline, please.
</blockquote></details>

Done.

---

*[src/app/app-loader.js, line 40 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFcBxKVDUbTHBKej3O1:-LFcRiFxA7H3gRinuzyo:b-896fix) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/7c03f6d61ea52cd1bfd6015fc2b8296ef6d6bab1/src/app/app-loader.js#L40)):*
<details><summary><i>Previously, barryytm (Barry Yung) wrote…</i></summary><blockquote>

Please insert a newline at the end of the file.
</blockquote></details>

Done.

---

*[src/content/styles/api/panel.css, line 97 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFcChxg1VTz7ekcCMu7:-LFcPq-yAsTA3K9b2s-L:b-xzdb3q) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/7c03f6d61ea52cd1bfd6015fc2b8296ef6d6bab1/src/content/styles/api/panel.css#L97)):*
<details><summary><i>Previously, barryytm (Barry Yung) wrote…</i></summary><blockquote>

Newline, please.
</blockquote></details>

Changed to SCSS file with newline -- done.

---

*[src/content/samples/api/new-panel-tests/panel/panel.css, line 97 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFcCjIK0VuYwtUgXj0i:-LFcRLJwELfzBgqc_yCO:blu3cqd) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/7c03f6d61ea52cd1bfd6015fc2b8296ef6d6bab1/src/content/samples/api/new-panel-tests/panel/panel.css#L97)):*
<details><summary><i>Previously, barryytm (Barry Yung) wrote…</i></summary><blockquote>

New line.Newline, please.
</blockquote></details>

File removed -- done. 

---


*Comments from [Reviewable](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804)*
<!-- Sent from Reviewable.io -->

barryytm commented:
Reviewed 5 of 8 files at r2.
Review status: 4 of 8 files reviewed, 23 unresolved discussions (waiting on @milesap, @dane-thomas, @ShrutiVellanki, @barryytm, @james-rae, and @JahidAhmed)

---

*[api/src/panel.ts, line 131 at r2](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFrJcpqCkQOSkYq6BLo:-LFrJcpqCkQOSkYq6BLp:bxvl20z) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/68e3e09681fd65ad5e242a7223d06dff0e4e6491/api/src/panel.ts#L131)):*
> ```XML
> 
>         //create panel components as HTMLElements
>         this._panel_contents = document.createElement("div");
> ```

Please use getters for attributes of panel.

---

*[api/src/panel.ts, line 168 at r2](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFrG8wG95Bhe7wYOCia:-LFrG8wG95Bhe7wYOCib:b6dfzvo) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/68e3e09681fd65ad5e242a7223d06dff0e4e6491/api/src/panel.ts#L168)):*
> ```XML
>     private windowResize(): void {
> 
>         let panel = this;
> ```

Instead of directly accessing the object from here.  Passing the object as a parameter (i.e. windowResize(panel)) might be a better way to do it.

---

*[api/src/panel.ts, line 172 at r2](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFrIwJz23t3NKsAIq1i:-LFrIwJz23t3NKsAIq1j:bxvl20z) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/68e3e09681fd65ad5e242a7223d06dff0e4e6491/api/src/panel.ts#L172)):*
> ```XML
>         $(window).resize(function () {
> 
>             let panelCoords = panel._panel_positions.panelCoords;
> ```

Please use getters for attributes of panel.

---

*[api/src/panel.ts, line 178 at r2](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFr0dU_6Pz9JfvVGkhT:-LFrGsMn5N7qRsEg1ID4:bxf5fh4) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/68e3e09681fd65ad5e242a7223d06dff0e4e6491/api/src/panel.ts#L178)):*
<details><summary><i>Previously, dane-thomas (Dane Thomas) wrote…</i></summary><blockquote>

I think you'd want the `undefined` check first so that it short circuits but I'm not sure if the `undefined` check is necessary in this case. Can someone else confirm?
</blockquote></details>

Please check the existence of `_map_width` unless it is always defined.

---


*Comments from [Reviewable](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804)*
<!-- Sent from Reviewable.io -->

ShrutiVellanki commented:
Review status: 2 of 8 files reviewed, 23 unresolved discussions (waiting on @barryytm, @milesap, @dane-thomas, @ShrutiVellanki, @james-rae, and @JahidAhmed)

---

*[api/src/map.ts, line 61 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFbis4GEzKPUR-CRywd:-LFrxzv2AqnyoFIk-9a2:b-896fix) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/7c03f6d61ea52cd1bfd6015fc2b8296ef6d6bab1/api/src/map.ts#L61)):*
<details><summary><i>Previously, ShrutiVellanki (Shruti Vellanki) wrote…</i></summary><blockquote>

comment added back; not pushing until PRs get pulled
</blockquote></details>

Done.

---

*[api/src/map.ts, line 62 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFbiyBUB0BbUrBbj98O:-LFrxyEw2V_16H49PCaw:b-896fix) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/8c0ef7fcdd3197222fb74f1596d932f67f33c17d/api/src/map.ts#L62)):*
<details><summary><i>Previously, ShrutiVellanki (Shruti Vellanki) wrote…</i></summary><blockquote>

comment added back; not pushing until PRs get pulled
</blockquote></details>

Done.

---

*[api/src/map.ts, line 67 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFbo6cADr4t3E4Tr6v6:-LFrFRL3DNyi7qAW6VH-:bx8ldil) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/7c03f6d61ea52cd1bfd6015fc2b8296ef6d6bab1/api/src/map.ts#L67)):*
<details><summary><i>Previously, milesap (Miles Petrov) wrote…</i></summary><blockquote>

What happens if there is more than one map on the page? Looks like it always uses the same element (the first map in DOM order).
</blockquote></details>

Done. Changed logic to retrieve inner shell for the specific map instance.

---

*[api/src/panel.ts, line 168 at r2](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFrG8wG95Bhe7wYOCia:-LFre9I2CcC6wmHpW97G:b-896fix) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/68e3e09681fd65ad5e242a7223d06dff0e4e6491/api/src/panel.ts#L168)):*
<details><summary><i>Previously, barryytm (Barry Yung) wrote…</i></summary><blockquote>

Instead of directly accessing the object from here.  Passing the object as a parameter (i.e. windowResize(panel)) might be a better way to do it.
</blockquote></details>

Done.

---

*[api/src/panel.ts, line 172 at r2](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFrIwJz23t3NKsAIq1i:-LFreD6b12UUQUBzHi2T:b8wys91) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/68e3e09681fd65ad5e242a7223d06dff0e4e6491/api/src/panel.ts#L172)):*
<details><summary><i>Previously, barryytm (Barry Yung) wrote…</i></summary><blockquote>

Please use getters for attributes of panel.
</blockquote></details>

Changed ._panel_positions to panelPositions getter. 

---

*[api/src/panel.ts, line 178 at r2](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFr0dU_6Pz9JfvVGkhT:-LFreUR99trY0zY-dsmi:b-896fix) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/68e3e09681fd65ad5e242a7223d06dff0e4e6491/api/src/panel.ts#L178)):*
<details><summary><i>Previously, barryytm (Barry Yung) wrote…</i></summary><blockquote>

Please check the existence of `_map_width` unless it is always defined.
</blockquote></details>

Done.

---

*[api/src/panel.ts, line 184 at r2](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFr11FH6-c_At6ax5DI:-LFreWUm1UMQzhXCcfJS:b-896fix) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/68e3e09681fd65ad5e242a7223d06dff0e4e6491/api/src/panel.ts#L184)):*
<details><summary><i>Previously, dane-thomas (Dane Thomas) wrote…</i></summary><blockquote>

Same as above 
</blockquote></details>

Done.

---

*[api/src/panel.ts, line 360 at r2](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFr2MWI0D_8VSeMyyen:-LFreXqh2TlBAyVxFC4s:b-896fix) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/68e3e09681fd65ad5e242a7223d06dff0e4e6491/api/src/panel.ts#L360)):*
<details><summary><i>Previously, dane-thomas (Dane Thomas) wrote…</i></summary><blockquote>

Maybe use every:
```js
panelCoords.every(coord => coord !== undefined)
```
</blockquote></details>

Done.

---

*[api/src/panel.ts, line 416 at r2](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFr3C2y7pWeZO8UTBcu:-LFreYzl9dXTdCaBwixU:b-896fix) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/68e3e09681fd65ad5e242a7223d06dff0e4e6491/api/src/panel.ts#L416)):*
<details><summary><i>Previously, dane-thomas (Dane Thomas) wrote…</i></summary><blockquote>

Same thing as above here
</blockquote></details>

Done.

---

*[api/src/panel.ts, line 590 at r2](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFr3Tbf6buBhAQDeAek:-LFreZtJ6ycXAKF7h631:b-896fix) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/68e3e09681fd65ad5e242a7223d06dff0e4e6491/api/src/panel.ts#L590)):*
<details><summary><i>Previously, dane-thomas (Dane Thomas) wrote…</i></summary><blockquote>

Same thing here. 
Also put this before the computation
</blockquote></details>

Done.

---

*[src/app/api-loader.ts, line 44 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFbom4o0hBGmi-O84yj:-LFrTEgIDrIPfqxSA4vL:bn5x3jo) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/7c03f6d61ea52cd1bfd6015fc2b8296ef6d6bab1/src/app/api-loader.ts#L44)):*
<details><summary><i>Previously, milesap (Miles Petrov) wrote…</i></summary><blockquote>

Since panels and their elements are closely tied to map instances, these classes should be accessible via a map instance (i.e. `RZ.mapInstances[0].Panel` instead of the global `RZ.Panel`. Then `PanelElem` and `Btn` should be accessible via a panel instance since they are tied to a specific panel. These could be statically accessed or only through an instance, whichever you think is best.
</blockquote></details>

Done. `createPanel` (in `Map`), `createPanelElem` and `createBtn`(in `Panel`) are now the only way to access them. Cannot be accessed through the globals `RZ.Panel`, `RZ.PanelElem`, `RZ.Btn`, etc.

---


*Comments from [Reviewable](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804)*
<!-- Sent from Reviewable.io -->

ShrutiVellanki commented:
Reviewed 2 of 8 files at r2, 1 of 5 files at r3, 6 of 6 files at r5.
Review status: all files reviewed, 23 unresolved discussions (waiting on @dane-thomas, @milesap, @ShrutiVellanki, @barryytm, @james-rae, and @JahidAhmed)

---

*[api/src/map.ts, line 57 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFbkp0rDPQiklBidVz1:-LFxazA9DHNAm6Fh5KCR:b-wexku5) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/7c03f6d61ea52cd1bfd6015fc2b8296ef6d6bab1/api/src/map.ts#L57)):*
<details><summary><i>Previously, milesap (Miles Petrov) wrote…</i></summary><blockquote>

A suggestion: move these properties to the map prototype and avoid using starting `_`'s :

```js
export interface Map {
    mapGrid: number[][];
    innerShell: HTMLElement;
    panelRegistry: Panel[];
}
```

Then you can remove these. This will keep the object more organised in dev consoles and will improve the readability of the map class.
</blockquote></details>

I have followed the format in this comment for `Panel`, `Map`, `PanelElem` and `Btn`  and placed class attributes (that I originally made private, now without  _) inside interfaces. Let me know if this was the correct way to go about it, of if I should explicitly be declaring them later as well (i.e: `Map.prototype.innerShell`)

---

*[api/src/panel.ts, line 63 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFbrsSc28jUlkVymVAe:-LFxbljC6TGTII8WZO5B:b-2tz816) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/7c03f6d61ea52cd1bfd6015fc2b8296ef6d6bab1/api/src/panel.ts#L63)):*
<details><summary><i>Previously, milesap (Miles Petrov) wrote…</i></summary><blockquote>

Move as many of these into the objects prototype (and remove starting `_`)
</blockquote></details>

Done. (see other comment)

---


*Comments from [Reviewable](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804)*
<!-- Sent from Reviewable.io -->

ShrutiVellanki commented:
Reviewed 1 of 1 files at r6, 3 of 3 files at r7.
Dismissed @barryytm from a discussion.
Review status: all files reviewed, 22 unresolved discussions (waiting on @dane-thomas, @milesap, @barryytm, @ShrutiVellanki, @james-rae, and @JahidAhmed)

---

*[api/src/panel.ts, line 310 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFcFkqh69jVcpVdTF0O:-LG0j29B71yWlXIDiSy4:b-896fix) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/7c03f6d61ea52cd1bfd6015fc2b8296ef6d6bab1/api/src/panel.ts#L310)):*
<details><summary><i>Previously, barryytm (Barry Yung) wrote…</i></summary><blockquote>

Put the svg code into a file or a constant?
</blockquote></details>

Done.

---

*[api/src/panel.ts, line 642 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFcFbBN4cLZV9ulFNZq:-LG0j4gv2JgOrXOfe7Gl:b-896fix) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/7c03f6d61ea52cd1bfd6015fc2b8296ef6d6bab1/api/src/panel.ts#L642)):*
<details><summary><i>Previously, barryytm (Barry Yung) wrote…</i></summary><blockquote>

Same as before.
</blockquote></details>

Done.

---

*[api/src/panel.ts, line 131 at r2](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFrJcpqCkQOSkYq6BLo:-LG0MIMtE02IpJypZbkO:b68ogtl) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/68e3e09681fd65ad5e242a7223d06dff0e4e6491/api/src/panel.ts#L131)):*
<details><summary><i>Previously, barryytm (Barry Yung) wrote…</i></summary><blockquote>

Please use getters for attributes of panel.
</blockquote></details>

I've changed what I have so far to use existing getters in most places. For `panel_contents`, `panel_controls`, etc. I don't want to implement getters because this would allow the API user to get the nodes of contents/controls when they should only be using `get content` and `get control` to access elements in array form. 

---


*Comments from [Reviewable](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804)*
<!-- Sent from Reviewable.io -->

milesap commented:
### Testing
URL: http://fgpv.cloudapp.net/demo/users/ShrutiVellanki/PanelApi/prod/samples/api/new-panel-tests/panel/panel-test.html

With the following commands:

```js
RZ.mapInstances[0].createPanel('p0');
RZ.mapInstances[0].panelRegistry[0].setPosition([5,5], [15,15]); //existing panel
RZ.mapInstances[0].panelRegistry[0].setMinPosition([7,7], [12,12]);
RZ.mapInstances[0].panelRegistry[1].setPosition([2,5], [8,18]); //panel p0
RZ.mapInstances[0].panelRegistry[1].setMinPosition([2,5], [6,18]);
RZ.mapInstances[0].panelRegistry[1].open();
```

### Expected result
Existing panel is re-positioned to`[7,5], [15,15]` and panel `p0` is  positioned at`[2,5], [6,18]`.

### Actual result
Panel  `p0` opened over the existing panel, existing panel did not resize. Closed existing panel, then when I tried to re-open an error `conflicting panels, panels cannot shrink any further to accomodate, panel closing.` was thrown.

---

Review status: all files reviewed, 22 unresolved discussions (waiting on @dane-thomas, @milesap, @barryytm, @james-rae, and @JahidAhmed)

---



*Comments from [Reviewable](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-:-LG1LrLTEXp7nxHNrvr6:bdzsfeu)*
<!-- Sent from Reviewable.io -->

milesap commented:
Review status: all files reviewed, 16 unresolved discussions (waiting on @dane-thomas, @milesap, @barryytm, @james-rae, and @JahidAhmed)

---

*[api/src/map.ts, line 57 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFbkp0rDPQiklBidVz1:-LG1PS2r0-5c_oDUgr7b:b-wum9iw) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/7c03f6d61ea52cd1bfd6015fc2b8296ef6d6bab1/api/src/map.ts#L57)):*
<details><summary><i>Previously, ShrutiVellanki (Shruti Vellanki) wrote…</i></summary><blockquote>

I have followed the format in this comment for `Panel`, `Map`, `PanelElem` and `Btn`  and placed class attributes (that I originally made private, now without  _) inside interfaces. Let me know if this was the correct way to go about it, of if I should explicitly be declaring them later as well (i.e: `Map.prototype.innerShell`)
</blockquote></details>

:+1: that is correct. You don't need to declare them explicitly on the prototype unless you want to set a default value (like an empty array for example).

---


*Comments from [Reviewable](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804)*
<!-- Sent from Reviewable.io -->

dane-thomas commented:
Review status: all files reviewed, 8 unresolved discussions (waiting on @ShrutiVellanki, @barryytm, @james-rae, and @JahidAhmed)

---

*[api/src/panel.ts, line 590 at r2](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFr3Tbf6buBhAQDeAek:-LG1PN8r09PE2pzxSt3M:beiazfg) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/68e3e09681fd65ad5e242a7223d06dff0e4e6491/api/src/panel.ts#L590)):*
<details><summary><i>Previously, ShrutiVellanki (Shruti Vellanki) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Maybe put the error checking before the rest of the function code. 

---


*Comments from [Reviewable](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804)*
<!-- Sent from Reviewable.io -->

milesap commented:
I've changed the target branch from `develop` to `panel-api`. Once you push a fix for my last test we'll likely merge into this branch so I can start working on angular integration. If any issues come up you can patch on that branch as well. We'll merge `panel-api` with `develop` once it is fully complete.

ShrutiVellanki commented:
Review status: 8 of 9 files reviewed, 8 unresolved discussions (waiting on @ShrutiVellanki, @barryytm, @james-rae, and @JahidAhmed)

---

*[api/src/panel.ts, line 590 at r2](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFr3Tbf6buBhAQDeAek:-LG5qjIf9eaJgyLKnGze:b-dqrqmv) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/68e3e09681fd65ad5e242a7223d06dff0e4e6491/api/src/panel.ts#L590)):*
<details><summary><i>Previously, dane-thomas (Dane Thomas) wrote…</i></summary><blockquote>

Maybe put the error checking before the rest of the function code. 
</blockquote></details>

Done. (please see latest revision)

---


*Comments from [Reviewable](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804)*
<!-- Sent from Reviewable.io -->

dane-thomas commented:
Review status: 8 of 9 files reviewed, 8 unresolved discussions (waiting on @ShrutiVellanki, @barryytm, @james-rae, and @JahidAhmed)

---

*[api/src/panel.ts, line 590 at r2](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-LFr3Tbf6buBhAQDeAek:-LG5tLks9gIQWSkPnrWM:b-8fx2u7) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/68e3e09681fd65ad5e242a7223d06dff0e4e6491/api/src/panel.ts#L590)):*
<details><summary><i>Previously, ShrutiVellanki (Shruti Vellanki) wrote…</i></summary><blockquote>

Done. (please see latest revision)
</blockquote></details>

:+1: 

---


*Comments from [Reviewable](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804)*
<!-- Sent from Reviewable.io -->

dane-thomas commented:
Reviewed 1 of 1 files at r8.
Review status: all files reviewed, 8 unresolved discussions (waiting on @ShrutiVellanki, @barryytm, @james-rae, and @JahidAhmed)

---



*Comments from [Reviewable](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804)*
<!-- Sent from Reviewable.io -->

milesap commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed, 8 unresolved discussions (waiting on @ShrutiVellanki, @barryytm, @james-rae, and @JahidAhmed)

---



*Comments from [Reviewable](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804#-:-LG62WX624HOHv5rJQcO:bnfp4nl)*
<!-- Sent from Reviewable.io -->

milesap commented:
Reviewed 2 of 8 files at r2, 1 of 5 files at r3, 3 of 6 files at r5, 2 of 3 files at r7, 1 of 1 files at r8.
Dismissed @barryytm from a discussion.
Review status: all files reviewed, 6 unresolved discussions (waiting on @barryytm, @james-rae, and @JahidAhmed)

---



*Comments from [Reviewable](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804)*
<!-- Sent from Reviewable.io -->

james-rae commented:
Reviewed 2 of 8 files at r2, 1 of 5 files at r3, 3 of 6 files at r5, 2 of 3 files at r7, 1 of 1 files at r8.
Review status: all files reviewed, 6 unresolved discussions (waiting on @barryytm and @JahidAhmed)

---



*Comments from [Reviewable](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804)*
<!-- Sent from Reviewable.io -->

barryytm commented:
Reviewed 1 of 8 files at r2, 2 of 5 files at r3, 3 of 6 files at r5, 2 of 3 files at r7, 1 of 1 files at r8.
Review status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @JahidAhmed)

---



*Comments from [Reviewable](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/2804)*
<!-- Sent from Reviewable.io -->
